src/EnergyPlus/Vectorize.hh

```diff
@@ -0,0 +1,34 @@
+#ifndef EnergyPlus_vectorize_hh_INCLUDED
+#define EnergyPlus_vectorize_hh_INCLUDED
+
+// EnergyPlus Vectorization Support
+
+// ObjexxFCL Headers
+#include <ObjexxFCL/vectorize.hh> // ASSUME_ALIGNED and RESTRICT macros
+
+// C++ Headers
+#include <cassert>
+#include <cstdint>
+
+// EnergyPlus headers
+#include "EnergyPlus.hh"
+
+namespace EnergyPlus {
+
+	int const VEC_LENGTH = VEC_ALIGN / sizeof(Real64);
+
+#define ROUND_TO_VEC_LENGTH(I) (((I) + ((VEC_LENGTH) - 1)) & ~((VEC_LENGTH) - 1))
+
+
+#define EXPLICIT_VECTORIZATION
+
+	// EXPLICIT_VECTORIZATION assumes that vectors are aligned,
+	// which is not a valid assumption on 32-bit builds
+#if defined(_WIN32) && !defined(_WIN64)
```

Comment:
@DeadParrot, thanks. HBIRE_vector_take2 is still not PR'ed because there
are few (3-4) outstanding "big" diffs, I can't tell if they are artificial
or not, and nobody has time to help me look right now obviously with the
holidays and the bug-fix release. Hopefully, this will get resolved after
the new year.

As for vectorization, please go ahead and give other parts of the code a
spin if you want. If the code is clean, it may vectorize on its own. I
don't know much about vectorization advisor, but I suspect that it tells
you *why* the compiler couldn't vectorize the loop so that you know what to
change? That sounds fun. In general, it's going to be one of four things:
i) memory ambiguities (can disambiguate using RESTRICT), ii) loop carried
dependences which require restructuring the loop (not that bad usually, I
did it in HBIRE), iii) non-conducive memory layout, e.g., accessing fields
in an array of objects (requires refactoring the data structure which could
have implications elsewhere, and this being EnergyPlus, you know there will
be :/), and iv) conditionals (the ResortedSurfaces branch has some code
that should allow us to eliminate the Surface(SurfNum).Class ==
SurfaceClass_Window conditionals, but wouldn't you know it? It generates
"big diffs"!).

May the force be with you! </sorry>

On Sun, Dec 20, 2015 at 8:42 PM, Stuart Mentzer <notifications@github.com>
wrote:

> In src/EnergyPlus/Vectorize.hh
> <https://github.com/NREL/EnergyPlus/pull/5278#discussion_r48112146>:
>
> > +
> > +// EnergyPlus headers
> > +#include "EnergyPlus.hh"
> > +
> > +namespace EnergyPlus {
> > +
> > +	int const VEC_LENGTH = VEC_ALIGN / sizeof(Real64);
> > +
> > +#define ROUND_TO_VEC_LENGTH(I) (((I) + ((VEC_LENGTH) - 1)) & ~((VEC_LENGTH) - 1))
> > +
> > +
> > +#define EXPLICIT_VECTORIZATION
> > +
> > +	// EXPLICIT_VECTORIZATION assumes that vectors are aligned,
> > +	// which is not a valid assumption on 32-bit builds
> > +#if defined(_WIN32) && !defined(_WIN64)
>
> Since I'm ready to PR the ObjexxFCL 4.1 release that improves on
> alignment/vectorization support (and other performance areas) it might make
> sense to rebase this branch on it.
>
> I learned some things about getting row-aligned 2D arrays to vectorize
> that could help here. I could also do some profiling and vectorization work
> on a couple of other promising loops using the new capabilities if that is
> of interest -- I'm itching to give the new Intel Vectorization Advisor a
> spin.
>
> I'm planning to post the ObjexxFCL PR as soon as the MArray_Elim is merged
> -- it may make sense to get those out of the way before we reopen the
> floodgates for general purpose commits.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NREL/EnergyPlus/pull/5278/files#r48112146>.
>